### PR TITLE
TreeKey sample app

### PR DESCRIPTION
**Pull request by loganj**
Built on #136 and #137.  Recommend reviewing/merging those first.

### CODE REVIEW

File: flow-sample-tree/src/main/java/flow/sample/tree/FlowServices.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      // Setting up the EditContactKey key means providing an editor for the contact.
+      // This editor can be shared among any keys that have the EditContactKey as parent/ancestor!
+      final String contactId = ((EditContactKey) key).contactId;
+      ContactsStorage storage = services.getService(CONTACTS_STORAGE);
+      Contact contact = storage.getContact(contactId);
+      services.bind(CONTACT_EDITOR, new ContactEditor(contact));
+    }
+  }
+
+  @Override public void tearDown(Services services) {
```
Comment (rjrjr) [inline flow-sample-tree/src/main/java/flow/sample/tree/FlowServices.java]: `tearDownServices` would better parallel `bindServices`

File: flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/edit/EditEmailView.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  ContactEditor editor;
+
+  public EditEmailView(Context context, AttributeSet attrs) {
+    super(context, attrs);
+    setOrientation(VERTICAL);
+    editor = Flow.getService(FlowServices.CONTACT_EDITOR, context);
+  }
+
+  @Override protected void onFinishInflate() {
```
Comment (rjrjr) [inline flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/edit/EditEmailView.java]: For consistency should use `onAttachedToWindow()`

File: flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/edit/EditNameView.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  private ContactEditor editor;
+
+  public EditNameView(Context context, AttributeSet attrs) {
+    super(context, attrs);
+    setOrientation(VERTICAL);
+    editor = Flow.getService(FlowServices.CONTACT_EDITOR, getContext());
+  }
+
+  @Override protected void onFinishInflate() {
```
Comment (rjrjr) [inline flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/edit/EditNameView.java]: `onAttachedToWindow`

File: flow/src/main/java/flow/ClassKey.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package flow;
+
+/**
+ * Convenience base class for keys.
```
Comment (rjrjr) [inline flow/src/main/java/flow/ClassKey.java]: Convenience base class for keys, makes all instances of the same subclass equal.

File: flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/ContactsUI.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package flow.sample.tree.ui.contacts;
+
+public final class ContactsUI {
```
Comment (rjrjr) [inline flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/ContactsUI.java]: Why not named with Key suffix, and why doesn't it extend ClassKey?

File: flow-sample-tree/src/main/java/flow/sample/tree/FlowServices.java (hunk 1 above)
Comment (loganj) [inline flow-sample-tree/src/main/java/flow/sample/tree/FlowServices.java]: :+1:

File: flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/edit/EditEmailView.java (hunk 2 above)
Comment (loganj) [inline flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/edit/EditEmailView.java]: yeah, just b/c of the click listener.  will change.

File: flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/ContactsUI.java (hunk 5 above)
Comment (loganj) [inline flow-sample-tree/src/main/java/flow/sample/tree/ui/contacts/ContactsUI.java]: :+1: didn't go back and fix this one

**Comment by loganj** [comment]
updated to address all feedback, build is green, UI tests pass. merging

**Pull request merged**